### ISSUE TITLE
Rewriting Scripts in Rust tracking issue

### ISSUE THREAD

**Issue by Maan2003** [fedimint/fedimint#2101]
We have a bunch of integration tests written in bash. They are flaky, hard to write and update. We want to re-write all these tests in Rust and minimize the amount of bash we depend on.

We've made 2 PRs in this project so far:
- https://github.com/fedimint/fedimint/pull/1899 Added ability to spawn daemons (e.g. `fedimintd` and `lnd` etc) via Rust
- https://github.com/fedimint/fedimint/pull/1910 Can write tests purely in rust. As a proof-of-concept, it re-wrote latency-test.sh: 
  - [Old bash version that was replaced](https://github.com/fedimint/fedimint/blob/5ef8158163c8cfa120b235e24c07999724bc33a0/scripts/latency-test.sh) 
  - [New rust version](https://github.com/fedimint/fedimint/blob/c77e5960255843890da809b3f3ebdcccc47e27e1/fedimint-bin-tests/src/main.rs#L417-L515) which is called [here](https://github.com/fedimint/fedimint/blob/c77e5960255843890da809b3f3ebdcccc47e27e1/scripts/latency-test.sh)

Work remaining (in order of importance):
- [ ] (1) Export `$FM_LOGS_DIR` in github actions. This helps us debug logs of specific daemons. See [upload-artifact](https://github.com/actions/upload-artifact)
- [x] Rewrite the rest of the bash integrations tests:
  - [x] (2a) [scripts/cli-test.sh](https://github.com/fedimint/fedimint/blob/master/scripts/cli-test.sh) (easier)
  - [x] (2b) [reconnect tests](https://github.com/fedimint/fedimint/blob/master/scripts/reconnect-test.sh) (harder b/c need to stop and start fedimintd)
  - [x] (2c) [upgrade tests](https://github.com/fedimint/fedimint/blob/master/scripts/upgrade-test.sh) (harder b/c need to stop and start fedimintd, but 2b has now demonstrated how to do it)
- [x] (3) Replace [scripts/setup-tests.sh](https://github.com/fedimint/fedimint/blob/master/scripts/setup-tests.sh) with a rust-spawned dev federation. For example, something like [this function](https://github.com/fedimint/fedimint/blob/57c55563680d40080c776cfa0f72b3ec7449255f/fedimint-bin-tests/src/main.rs#L557-L570). Perhaps rename to dev_federation or something. blocked on (2b) (2c)
- [x] (4) Spawn `esplora` in fedimint-bin-tests. This functionality exists in [fedimint-testing/src/bin/fixtures.rs](https://github.com/fedimint/fedimint/blob/89f83f3eb07dbdac21b6eba89f7f5764fcbe6212/fedimint-testing/src/bin/fixtures.rs#L243-L266)
- [x] (5) Spawn `electrs` in fedimint-bin-tests. This functionality exists in [fedimint-testing/src/bin/fixtures.rs](https://github.com/fedimint/fedimint/blob/89f83f3eb07dbdac21b6eba89f7f5764fcbe6212/fedimint-testing/src/bin/fixtures.rs#L224-L241)
- [x] (6) Don't set any environment variables in [scripts/build.sh](https://github.com/fedimint/fedimint/blob/master/scripts/build.sh). Just run `cargo build ...`. Instead, set environment variables in [dev_fed](https://github.com/fedimint/fedimint/blob/57c55563680d40080c776cfa0f72b3ec7449255f/fedimint-bin-tests/src/main.rs#L526-L555) Rust function.
- [x] (7) Once nothing depends on it, remove [scripts/lib.sh](https://github.com/fedimint/fedimint/blob/master/scripts/lib.sh)
- [x] (8) Once nothing depends on it, remove [fedimint-testing/src/bin/fixtures.rs](https://github.com/fedimint/fedimint/blob/master/fedimint-testing/src/bin/fixtures.rs)
- [x] (9) https://github.com/fedimint/fedimint/issues/2137
- [ ] (10) Ability to run dkg with n peers and run `fedimintd` with fewer than n peers 
- [x] (11) Ability to run `fedimintd` without running DKG beforehand. This will allow us to replace [scripts/run-ui.sh](https://github.com/fedimint/fedimint/blob/master/scripts/run-ui.sh) which is useful to test setup UI
- [x] Use `PATH` instead of `FM_BIN_DIR`
- [x] When any `cmd!` fails, log the the entire command that was run
- [x] devimint should peg-in user and gateways for mprocs and tmux developer environments. [Stop using pegin.sh](https://github.com/fedimint/fedimint/blob/a8b6b78328081f000794b5020f6b93c60a8ee7ef/scripts/mprocs-user-shell.sh#L19-L26) and then delete it.
- [x] Write `distributedgen` to a log file. Probably the same one that `fedimintd` uses. This way errors will show up visually in `mprocs`.
- [ ] When `devimint tmuxinator` is running, we should have a command to start / stop the fedimintd servers.
- [x] `include!` [daemon config files](https://github.com/fedimint/fedimint/tree/master/misc/test) so external projects don't need them.
  - [ ] Eventually an option to override these default configs

If you'd like to tackle one of these issues, just post "I'll try (\<n\>)". If multiple people want to try the same thing, try to collaborate in discord. Make [draft pull requests](https://egghead.io/lessons/github-open-a-github-draft-pull-request) early for feedback. @Maan2003 can help you!

**Comment by Maan2003** [comment]
will re add good first issue ater latency tests are fixed.

**Comment by ayush-ranjan-official** [comment]
@Maan2003 I'll try (4)

**Comment by gotlougit** [comment]
I'd like to work on (5)

**Comment by GetPsyched** [comment]
I'd like to work on (2a)

**Comment by Maan2003** [comment]
@gotlougit @ayush-ranjan-official (4) and (5) are very similar, feel free to collaborate and do it in one PR.

**Comment by oleonardolima** [comment]
@Maan2003 I'd like to work on (3)

**Comment by Maan2003** [comment]
FYI you don't need to ping me to say that. You just need to make sure that there is no duplicated work.

But feel free to ping me here or on discord if you get stuck somewhere.

**Comment by Maan2003** [comment]
@oleonardolima updated description of (3)

**Comment by Maan2003** [comment]
@oleonardolima that update was incorrect, Sorry. updated again

**Comment by debjit-bw** [comment]
I want to take up (6)

**Comment by ayush-ranjan-official** [comment]
> @Maan2003 I'll try (4) I guess, (4) is solved, so I would like to try (1)

**Comment by oleonardolima** [comment]
> @oleonardolima that update was incorrect, Sorry. updated again

No problem, as it's blocked on (2b) and (2c), I'll work on (2b) first.

**Comment by dhruv-1001** [comment]
Hey @oleonardolima I was working on (2b), sorry I forgot to mention it here and started working on it. Do you mind if I continue with (2b)? (I'm ready to push a draft PR)

**Comment by Maan2003** [comment]
There is no harm in opening a draft PR. 
You should collaborate to get the rest of the work done.

**Comment by Maan2003** [comment]
Added a reference for 1

**Comment by justinmoon** [comment]
Now that we've completed 2b, 2c should be easier because there are hooks to start and stop fedimintd

**Comment by dhruv-1001** [comment]
If no one is working on 2c, i'll take this one.

**Comment by justinmoon** [comment]
> If no one is working on 2c, i'll take this one.

Go for it!

**Comment by jkitman** [comment]
@dhruv-1001 Don't worry about (2c) I'm going to test that one in another way (I'm changing how the upgrade API works).

**Comment by Maan2003** [comment]
I am working (6).

**Comment by Maan2003** [comment]
I am working on (3).

**Comment by Maan2003** [comment]
> devimint should peg-in user and gateways for mprocs and tmux developer environments. [Stop using pegin.sh](https://github.com/fedimint/fedimint/blob/a8b6b78328081f000794b5020f6b93c60a8ee7ef/scripts/mprocs-user-shell.sh#L19-L26) and then delete it.

we can't delete scripts/pegin.sh until we have replacement for it that can invoked from the user shell.

**Comment by justinmoon** [comment]
> we can't delete scripts/pegin.sh until we have replacement for it that can invoked from the user shell.

Good point. Let's wait for that until @elsirion finishes the wallet commands for client-ng. That's what he's working on now.

**Comment by justinmoon** [comment]
But the user shell wouldn't need to do this. `dev_fed` could do this.

**Comment by techsavvyash** [comment]
i'll try (17 (`include!` daemon config files)

**Comment by justinmoon** [comment]
@techsavvyash: Great. Feel free to just make a draft PR once you've got it working with 1 daemon.

**Comment by justinmoon** [comment]
Closing this because the initial project is complete. Might add another issue with devimint improvement ideas.
